In [1]:
import pandas as pd
import numpy as np
from datetime import datetime
from datetime import date
import matplotlib.pyplot as plt
import seaborn as sns
import requests
pd.options.display.max_rows = 5000 
pd.options.display.max_columns = 200

### Part 1
达到第一个目标收益比所用的天数统计，比如,达到20%盈利用了多少天

In [4]:
def days_analysis(excel_name, p1_date, p1_price,p1_wprice):
#     读取数据,回复index，简化col names, 变量化配售日期，配售价格和权证价格
    df = pd.read_csv(r'/Users/paxton615/Desktop/Stock_Price_csv/si_stocks/{}.csv'.format(excel_name))
    df.drop('Unnamed: 0',axis=1, inplace=True)
    df.columns = ['dates', 'close', 'volume']
    p1_date = p1_date
    p1_price = p1_price
    p1_wprice = p1_wprice
#     从配售日期开始截取数据，将数据从远期往近期排列
    df = df[df['dates']>=p1_date]
    df = df[::-1].reset_index(drop=True)
#     分别计算股价收益，权证收益和总收益（自然数，而非百分比）
    df['share_rtn' ]= [(x-p1_price) if (x-p1_price)>0 else 0 for x in df['close']]
#     权证价格，首先判断不为0，权证收益小于0则等于0，权证收益大于零则按正常记录
    new = []
    for x in df['close']:
        if p1_wprice == 0:
            new.append(0)
        elif (x-p1_wprice)<=0 :
            new.append(0)
        else :
            new.append(x-p1_wprice)
    df['warrant_rtn']=new
        
    df['total_rtn'] = df['share_rtn' ] + df['warrant_rtn']
#     读取数据中最高收盘价，最高收盘价日期，和最高收益（自然数）,最高ROI（小数百分比）
    trade_high = df['close'].max()
#     trade_high_date = (df.loc[df['close']==df['close'].max(), 'dates'].item())
    highest_rtn = df['total_rtn'].max()
    highest_roi = np.round(df['total_rtn'].max()/p1_price,4)
    
#     新建一个df，每行只保留综合数据
    placing = pd.DataFrame({'placing':excel_name, 'placing_date':p1_date, 'placing_price':p1_price, 'warrant_price':p1_wprice ,
                          'trade_high':trade_high, 'highest_rtn':highest_rtn, 'highest_roi':highest_roi},index=[0]) #'trade_high_date':trade_high_date, 代码有问题
    
    
    placing['all_trade_days'] = len(df['close'])
    
#     达到目标盈利，用了多少天
    count = 1
    for i in df['total_rtn']:
        if i/p1_price <= 0.2:
            count +=1
        else:
            break
    placing['days_to_20%'] = [count if count!=placing['all_trade_days'][0]+1 else 0]
    
    count = 1
    for i in df['total_rtn']:
        if i/p1_price <= 0.5:
            count +=1
        else:
            break
    placing['days_to_50%'] = [count if count!=placing['all_trade_days'][0]+1 else 0]
    
    count = 1
    for i in df['total_rtn']:
        if i/p1_price <= 0.7:
            count +=1
        else:
            break
    placing['days_to_70%'] = [count if count!=placing['all_trade_days'][0]+1 else 0]
    
    count = 1
    for i in df['total_rtn']:
        if i/p1_price <= 1:
            count +=1
        else:
            break
    placing['days_to_100%'] = [count if count!=placing['all_trade_days'][0]+1 else 0]
    
    count = 1
    for i in df['total_rtn']:
        if i/p1_price <= 1.5:
            count +=1
        else:
            break
    placing['days_to_150%'] = [count if count!=placing['all_trade_days'][0]+1 else 0]
    
    count = 1
    for i in df['total_rtn']:
        if i/p1_price <= 1.75:
            count +=1
        else:
            break
    placing['days_to_175%'] = [count if count!=placing['all_trade_days'][0]+1 else 0]
    
    
    count = 1
    for i in df['total_rtn']:
        if i/p1_price <= 2:
            count +=1
        else:
            break
    placing['days_to_200%'] = [count if count!=placing['all_trade_days'][0]+1 else 0]
    
    count = 1
    for i in df['total_rtn']:
        if i/p1_price <= 3:
            count +=1
        else:
            break
    placing['days_to_300%'] = [count if count!=placing['all_trade_days'][0]+1 else 0]
    
    count = 1
    for i in df['total_rtn']:
        if i/p1_price <= 4:
            count +=1
        else:
            break
    placing['days_to_400%'] = [count if count!=placing['all_trade_days'][0]+1 else 0]
    
    count = 1
    for i in df['total_rtn']:
        if i/p1_price <= 5:
            count +=1
        else:
            break
    placing['days_to_500%'] = [count if count!=placing['all_trade_days'][0]+1 else 0]
    
    return(placing)

In [5]:
cobr_p1 = days_analysis('cobr','2020-01-01', 1.00, 2.00)

cobr_p2 = days_analysis('cobr','2020-09-01', 2.30, 3.00)

palm_p1 = days_analysis('palm', '2020-01-01', 6.0, 12.0)

palm_p2 = days_analysis('palm', '2020-07-01', 6.5, 0.0)

palm_p3 = days_analysis('palm', '2020-12-01', 10.0, 0.0)

amte_p1 = days_analysis('amte','2019-07-01', 73.5, 0.0)

amte_p2 = days_analysis('amte','2020-04-01', 73.5, 0.0)

amte_p3 = days_analysis('amte','2021-03-01', 175.0, 0.0)

ecr_p1 = days_analysis('ecr', '2020-04-01', 0.5, 1.0)

ecr_p2 = days_analysis('ecr', '2020-07-01', 1.3, 2.1)

kat_p1 = days_analysis('kat','2020-07-01',1.5, 3.0)

kat_p2 = days_analysis('kat','2020-02-01',2.0, 3.0)

kat_p3 = days_analysis('kat','2021-01-01',2.0, 3.0)

irr_p1 = days_analysis('irr', '2020-05-01', 7.0, 12.0)

irr_p2 = days_analysis('irr', '2021-04-01', 20.0, 0.0)

mnrg_p1 = days_analysis('mnrg','2020-07-01', 0.6, 0.0)

mnrg_p2 = days_analysis('mnrg','2021-03-01', 0.6, 1.0)

ccz_p1 = days_analysis('ccz', '2020-08-01', 1.7, 2.8)

ccz_p2 = days_analysis('ccz', '2021-06-01', 2.3, 4.4)

imm_p1 = days_analysis('imm', '2020-09-01', 11.0, 0)

avo_p1 = days_analysis('avo', '2020-11-01', 30, 0.0 )

avo_p2 = days_analysis('avo', '2021-01-01', 40, 0.0 )

cgh_p1 = days_analysis('cgh', '2021-02-01',26.0, 0.0)

knb_p1 = days_analysis('knb', '2020-04-01',0.0005, 0.0)

knb_p2 = days_analysis('knb', '2021-02-01',6.5, 0.0)

thr_p1 = days_analysis('thr', '2020-09-01', 0.6, 1.0)

kav_p1 = days_analysis('kav','2020-04-01', 0.8, 1.0)

kav_p2 = days_analysis('kav','2020-12-01', 2.75, 4.25)

acp_p1 = days_analysis('acp', '2019-09-01', 1.8, 3.0)

In [6]:
frames = [cobr_p1, cobr_p2, palm_p1,palm_p2,palm_p3, amte_p1,amte_p2,amte_p3,
         ecr_p1, ecr_p2,kat_p1, kat_p2, kat_p3, irr_p1, irr_p2, mnrg_p1, mnrg_p2,
         ccz_p1, ccz_p2, imm_p1, avo_p1, avo_p2, cgh_p1, knb_p1, knb_p2, thr_p1,
         kav_p1, kav_p2, acp_p1]
days_analysis = pd.concat(frames)
days_analysis

placing placing_date  placing_price  warrant_price  trade_high  highest_rtn  \
0    cobr   2020-01-01         1.0000           2.00       3.500       4.0000   
0    cobr   2020-09-01         2.3000           3.00       3.500       1.7000   
0    palm   2020-01-01         6.0000          12.00      17.000      16.0000   
0    palm   2020-07-01         6.5000           0.00      17.000      10.5000   
0    palm   2020-12-01        10.0000           0.00      17.000       7.0000   
0    amte   2019-07-01        73.5000           0.00     327.000     253.5000   
0    amte   2020-04-01        73.5000           0.00     327.000     253.5000   
0    amte   2021-03-01       175.0000           0.00     327.000     152.0000   
0     ecr   2020-04-01         0.5000           1.00       4.250       7.0000   
0     ecr   2020-07-01         1.3000           2.10       4.250       5.1000   
0     kat   2020-07-01         1.5000           3.00       3.000       1.5000   
0     kat   2020-02-01         2.0000           3.00       4.350       3.7000   
0     kat   2021-01-01         2.0000           3.00       2.480       0.4800   
0     irr   2020-05-01         7.0000          12.00      26.500      34.0000   
0     irr   2021-04-01        20.0000           0.00      24.400       4.4000   
0    mnrg   2020-07-01         0.6000           0.00       1.300       0.7000   
0    mnrg   2021-03-01         0.6000           1.00       1.150       0.7000   
0     ccz   2020-08-01         1.7000           2.80       4.400       4.3000   
0     ccz   2021-06-01         2.3000           4.40       2.875       0.5750   
0     imm   2020-09-01        11.0000           0.00      16.250       5.2500   
0     avo   2020-11-01        30.0000           0.00      47.500      17.5000   
0     avo   2021-01-01        40.0000           0.00      47.500       7.5000   
0     cgh   2021-02-01        26.0000           0.00      30.900       4.9000   
0     knb   2020-04-01         0.0005           0.00      40.800      40.7995   
0     knb   2021-02-01         6.5000           0.00      40.800      34.3000   
0     thr   2020-09-01         0.6000           1.00       1.275       0.9500   
0     kav   2020-04-01         0.8000           1.00       6.200      10.6000   
0     kav   2020-12-01         2.7500           4.25       6.200       5.4000   
0     acp   2019-09-01         1.8000           3.00       6.250       7.7000   

   highest_roi  all_trade_days  days_to_20%  days_to_50%  days_to_70%  \
0       4.0000             459            1           24           24   
0       0.7391             301           35           36           36   
0       2.6667             464          101          143          144   
0       1.6154             344           21           23           26   
0       0.7000             236           10           22            0   
0       3.4490             166            1            1            1   
0       3.4490             166            1            1            1   
0       0.8686             166            1            2            4   
0      14.0000             405            9           10           11   
0       3.9231             344            8           19           19   
0       1.0000             344            2           28           29   
0       1.8500             447            5            6            6   
0       0.2400             215           13            0            0   
0       4.8571             385            1           42           44   
0       0.2200             152           67            0            0   
0       1.1667             344           72           77           77   
0       1.1667             175           11           11           12   
0       2.5294             320            1           11           53   
0       0.2500             113           31            0            0   
0       0.4773             301            1            0            0   
0       0.5833             257        

- MAKE A COPY OF DATA

In [7]:
wait_days = days_analysis

In [8]:
wait_days.describe()

placing_price  warrant_price  trade_high  highest_rtn   highest_roi  \
count      29.000000      29.000000   29.000000    29.000000     29.000000   
mean       17.474155       2.018966   46.952759    30.881190   2816.395828   
std        36.144623       3.119421   97.892595    67.915406  15152.047070   
min         0.000500       0.000000    1.150000     0.480000      0.187500   
25%         1.300000       0.000000    3.500000     3.700000      0.700000   
50%         2.300000       1.000000    6.250000     5.400000      1.615400   
75%        11.000000       3.000000   30.900000    16.000000      3.923100   
max       175.000000      12.000000  327.000000   253.500000  81599.000000   

       all_trade_days  days_to_20%  days_to_50%  days_to_70%  days_to_100%  \
count       29.000000    29.000000    29.000000    29.000000     29.000000   
mean       288.896552    18.068966    23.862069    24.206897     30.310345   
std        113.537705    25.334802    34.472271    35.972588     43.882232   
min        113.000000     0.000000     0.000000     0.000000      0.000000   
25%        184.000000     1.000000     1.000000     0.000000      0.000000   
50%        301.000000     9.000000    11.000000    11.000000      1.000000   
75%        344.000000    21.000000    28.000000    29.000000     64.000000   
max        554.000000   101.000000   143.000000   144.000000    146.000000   

       days_to_150%  days_to_175%  days_to_200%  days_to_300%  days_to_400%  \
count     29.000000     29.000000     29.000000     29.000000     29.000000   
mean      40.241379     37.551724     33.137931     27.655172     30.793103   
std       60.613916     65.352116     63.464683     63.027361    101.731572   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        1.000000      0.000000      0.000000      0.000000      0.000000   
75%       77.000000     65.000000     66.000000      4.000000      0.000000   
max      204.000000    249.000000    250.000000    254.000000    514.000000   

       days_to_500%  
count     29.000000  
mean       7.620690  
std       29.133209  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        0.000000  
max      138.000000

#### 剔除0后，到达各个回报率所用天数descrebe( )

In [19]:
wait_days[wait_days['days_to_20%']!=0]['days_to_20%'].describe()

count     27.000000
mean      19.407407
std       25.773062
min        1.000000
25%        1.000000
50%       10.000000
75%       26.000000
max      101.000000
Name: days_to_20%, dtype: float64

In [20]:
wait_days[wait_days['days_to_50%']!=0]['days_to_50%'].describe()

count     23.000000
mean      30.086957
std       36.285375
min        1.000000
25%        8.000000
50%       20.000000
75%       38.500000
max      143.000000
Name: days_to_50%, dtype: float64

In [21]:
wait_days[wait_days['days_to_70%']!=0]['days_to_70%'].describe()

count     21.000000
mean      33.428571
std       38.574048
min        1.000000
25%        6.000000
50%       20.000000
75%       44.000000
max      144.000000
Name: days_to_70%, dtype: float64

In [22]:
wait_days[wait_days['days_to_100%']!=0]['days_to_100%'].describe()

count     18.000000
mean      48.833333
std       47.052224
min        1.000000
25%        7.500000
50%       36.000000
75%       77.500000
max      146.000000
Name: days_to_100%, dtype: float64

In [23]:
wait_days[wait_days['days_to_200%']!=0]['days_to_200%'].describe()

count     12.000000
mean      80.083333
std       78.429422
min        1.000000
25%        1.750000
50%       72.000000
75%      111.750000
max      250.000000
Name: days_to_200%, dtype: float64

- 通过统计我们可以看出，

达到20%盈利标准，平均18天，最长101天，最短当天；

达到50%盈利标准，平均24天，最长143天，最短当天；

达到70%盈利标准，平均24天，最长144天，最短当天；

达到100%盈利标准，平均30天，最长146天，最短当天；

达到200%盈利标准，平均33天，最长250天，最短当天；


有多少支股票有20%的盈利，占所有placing次数的百分比

In [9]:
print(len(wait_days[wait_days['days_to_20%']!=0]),'\n',np.round(len(wait_days[wait_days['days_to_20%']!=0])/29,2))

27 
 0.93


有多少支股票有50%的盈利，占所有placing次数的百分比

In [10]:
print(len(wait_days[wait_days['days_to_50%']!=0]),'\n',np.round(len(wait_days[wait_days['days_to_50%']!=0])/29,2))

23 
 0.79


有多少支股票有100%的盈利，占所有placing次数的百分比

In [11]:
print(len(wait_days[wait_days['days_to_100%']!=0]),'\n',np.round(len(wait_days[wait_days['days_to_100%']!=0])/29,2))

18 
 0.62


有多少支股票有200%的盈利，占所有placing次数的百分比

In [12]:
print(len(wait_days[wait_days['days_to_200%']!=0]),'\n',np.round(len(wait_days[wait_days['days_to_200%']!=0])/29,2))

12 
 0.41


有多少支股票有300%的盈利，占所有placing次数的百分比

In [13]:
print(len(wait_days[wait_days['days_to_300%']!=0]),'\n',np.round(len(wait_days[wait_days['days_to_300%']!=0])/29,2))

10 
 0.34


有多少支股票有400%的盈利，占所有placing次数的百分比

In [14]:
print(len(wait_days[wait_days['days_to_400%']!=0]),'\n',np.round(len(wait_days[wait_days['days_to_400%']!=0])/29,2))

6 
 0.21


有多少支股票有400%的盈利，占所有placing次数的百分比

In [15]:
print(len(wait_days[wait_days['days_to_500%']!=0]),'\n',np.round(len(wait_days[wait_days['days_to_500%']!=0])/29,2))

4 
 0.14


In [16]:
# 以上function是通过以下代码推算出来的


# df = pd.read_csv(r'/Users/paxton615/Desktop/Stock_Price_csv/cobr.csv')
# df.drop('Unnamed: 0',axis=1, inplace=True)
# df.columns = ['dates', 'close', 'volume']

# p1_date = '2020-01-01'
# p1_price = 1
# p1_wprice = 2

# df.close.plot();
# df.total_rtn.plot();

# df = df[df['dates']>=p1_date]
# df = df[::-1].reset_index(drop=True)
# df['share_rtn' ]= [(x-p1_price) if (x-p1_price)>0 else 0 for x in df['close']]

# new = []
# for x in df['close']:
#     if p1_wprice == 0:
#         new.append(0)
#     elif (x-p1_wprice)<=0 :
#         new.append(0)
#     else :
#         new.append(x-p1_wprice)
        
# df['warrant_rtn']=new
        
# df['total_rtn'] = df['share_rtn' ] + df['warrant_rtn'] 

# df['all_trade_days'] = len(df['close'])
    

# count = 1
# for i in df['total_rtn']:
#     if i/p1_price <= 0.2  :
#         count +=1
#     else:
#         break

# count

# [count if count != (df['all_trade_days'][0]+1) else 0]

# (df['all_trade_days'][0]+1)
